src: add detectEBPFSupport API & eBPF loader

RafaelGSS opened the pull request:
This commit introduces a new initialization in the startup queue: initializeEbpf()

The startup attempts to load into the kernel all ebpf programs in the allowed_programs list.

---

Depends on: #320 

```console
$ sudo NODE_DEBUG_NATIVE=NSOLID_EBPF NSOLID_LOAD_EBPF=1 ./nsolid
```

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Summary by CodeRabbit

- **New Features**
  - Added comprehensive BPF headers, core helpers, tracing macros, and utilities for advanced eBPF program development.
  - Introduced extensive libbpf documentation covering API usage, naming conventions, program types, and build instructions.
  - Added robust BPF Type Format (BTF) support with creation, parsing, relocation, and dumping capabilities.
  - Implemented ELF parsing utilities for symbol resolution to support uprobes.
  - Included kernel feature detection probes to enable adaptive behavior based on kernel capabilities.
  - Added new CI and build scripts for Debian, Ubuntu, Travis, fuzzing, Coverity, and kernel synchronization automation.
  - Introduced BPF CO-RE relocatable access macros for portable kernel data structure reads.
  - Provided new BPF selftest runners with allowlist and denylist configurations for various kernel versions.
  - Added support for AF_XDP socket interfaces and netdev attributes.
  - Introduced traffic control (tc) classifier and scheduler definitions for extended networking support.

- **Bug Fixes**
  - Fixed BPF selftests and kernel config compatibility issues, including relocation of speculation mitigations to generic config.

- **Chores**
  - Added licensing files (LGPL-2.1 and BSD-2-Clause) and contributor mailmap for consistent attribution.
  - Enhanced build system with Makefiles, GYP configs, and pkg-config integration.
  - Added gitignore files to exclude build artifacts and documentation outputs.
  - Added helper scripts for CI log formatting and test automation.
  - Integrated static analysis and fuzzing infrastructure.

- **Documentation**
  - Added detailed libbpf API references, BPF CO-RE programming guides, and BTF format explanations.
  - Provided README and synchronization instructions for libbpf source maintenance.
  - Added Sphinx and Doxygen documentation setup with themed output and API cross-referencing.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

RafaelGSS committed:
src: add nsolid.detectEBPFSupport()

This API should return an object showing
all kernel capabilities related to ebpf

Signed-off-by: RafaelGSS <rafael.nunu@hotmail.com>

RafaelGSS commented on node.gyp:
```diff
@@ -1125,6 +1125,13 @@
+        [ 'OS=="linux"', {
+          'link_settings': {
+            'libraries': [
+              '-lbpf',
```
Not sure if that's desired as we won't have a conditional if the system doesn't have libbpf installed.

RafaelGSS commented on src/nsolid/nsolid_bpf-inl.h:
```diff
@@ -60,6 +72,64 @@ struct EBPFSupportInfo {
+    return EbpfLoadStatus::LOAD_ERROR;
+  }
+
+  // bpf_object__close(obj);
```
Should we store all bpf programs and close them on process shutdown?

RafaelGSS commented on src/nsolid/nsolid_bpf-inl.h:
```diff
@@ -270,13 +340,41 @@ inline EBPFSupportInfo detectEBPFSupport() {
   return info;
 }
 
+inline void loadEbpfPrograms() {
```
This will be removed after PR approval.

RafaelGSS commented on src/env.cc:
```diff
@@ -9,7 +9,9 @@
 #include "node_context_data.h"
 #include "node_contextify.h"
 #include "node_errors.h"
-#include "node_internals.h"
```
revert it

santigimeno commented on src/nsolid/nsolid_bpf.h:
```diff
@@ -46,10 +54,38 @@ inline void cap_free(cap_t) {}
+  EbpfLoader();
+  ~EbpfLoader();
+
+  EbpfLoadStatus LoadProgram(const std::string& program_name);
+
+  // Clean up all loaded BPF objects
+  static void CleanupAllBpfObjects();
+
+ private:
+  const std::unordered_set<std::string> allowed_programs_;
+  const std::string ebpf_program_path_;
+
+  // Static vector to keep track of all loaded BPF objects
+#if HAVE_LIBBPF
+  static std::vector<struct ::bpf_object*> loaded_objects_;
+#endif
+};
```
Why making `loaded_objects_` static (along with the `CleanupAllBpfObjects()` method?

RafaelGSS replied:
This way, we have full control of all loaded objects for a proper cleanup. Otherwise, on process shutdown, we'll need to get all loader instances and access their loaded objects (to simplify lifecycle).

RafaelGSS replied:
For example, if multiple Envs are created (worker threads), we would need to deal with loading ePBF programs twice. Adding it to node_main_instance (and having a static method), we ensure we just have one place with all loaded programs.

santigimeno replied:
But, it doesn't need to be static: make it part of EnvList and access the method you want (for cleanup or whatever) via it. It can be a normal class

santigimeno replied:
EnvList is our class for everything global: not linked to a specific environment.

RafaelGSS replied:
Oh, right. I thought we had one EnvList per Env.

santigimeno commented on src/nsolid/nsolid_bpf.cc:
```diff
@@ -1,9 +1,78 @@
 namespace node {
 namespace nsolid {
 
+#if HAVE_LIBBPF
+std::vector<struct ::bpf_object*> EbpfLoader::loaded_objects_;
+static std::mutex bpf_objects_mutex;
+#endif
```
Why don't make the mutex part of the `EbpfLoader` class: we should only be accessing loaded_objects_ via `EbpfLoader` methods.

RafaelGSS replied:
Actually, I think the mutex is not necessary anymore. A deadlock might only occur if the process is shutting down and we attempt to load a new ebpf program. At this point, this won't happen.

RafaelGSS replied:
Just removed.

santigimeno commented on src/nsolid/nsolid_api.cc:
```diff
@@ -2658,6 +2663,42 @@ uint32_t calculateKernelVersion() {
 }
 #endif
 
+void loadEbpfPrograms() {
+  EbpfLoader loader;
+  Debug(DebugCategory::NSOLID_EBPF, "Attempting to load eBPF programs...\n");
+
+  EbpfLoadStatus status = loader.LoadProgram("hello_world");
+
+  switch (status) {
+    case EbpfLoadStatus::SUCCESS:
+      Debug(DebugCategory::NSOLID_EBPF,
+            "eBPF program 'hello_world' loaded successfully.\n");
+      break;
+    case EbpfLoadStatus::FILE_NOT_FOUND:
+      Debug(DebugCategory::NSOLID_EBPF,
+            "Error - eBPF program 'hello_world' file not found.\n");
+      break;
+    case EbpfLoadStatus::NOT_ALLOWED:
+      Debug(DebugCategory::NSOLID_EBPF,
+            "Error - eBPF program 'hello_world' is not in the allowed list.\n");
+      break;
+    case EbpfLoadStatus::LOAD_ERROR:
+      Debug(DebugCategory::NSOLID_EBPF,
+            "Error - Failed to load eBPF program 'hello_world'.\n");
+      break;
+    case EbpfLoadStatus::NOT_SUPPORTED:
+      Debug(DebugCategory::NSOLID_EBPF,
+            "Failed to load: either eBPF is not supported or missing required "
+            "privileges (e.g., not running with sudo)\n");
+      break;
+    default:
+      Debug(
+          DebugCategory::NSOLID_EBPF,
+          "Error - Unknown status while loading eBPF program 'hello_world'.\n");
+      break;
+  }
+}
+
```
This code seems misplaced here. It should be part of `EnvList` interface, imho

RafaelGSS replied:
I was assuming we'll need to load ebpf programs during runtime (not only during initialization). That's why I added to nsolid_api, to have a JS API for it.

santigimeno commented on src/nsolid/nsolid_api.h:
```diff
@@ -970,6 +970,7 @@ inline uv_loop_t* EnvList::thread_loop() {
 
 uint32_t calculateKernelVersion();
 
+void loadEbpfPrograms();
```
Move this method into `EnvList`.

santigimeno commented on src/env.h:
```diff
@@ -652,6 +652,7 @@ class Environment final : public MemoryRetainer {
   void RunDeserializeRequests();
   // Should be called before InitializeInspector()
   void InitializeDiagnostics();
+  void InitializeEbpf();
```
We should try to avoid growing Node.js internal interfaces as much as possible. If we were to add a new API linked to an Environment, we should do it by adding the api to `EnvInst`. In this case I don't think the initialization belongs to the environment, because it's a "global" (only to be called once) thing, so I would move it to EnvList.

santigimeno commented on src/node_main_instance.cc:
```diff
@@ -99,6 +99,8 @@ ExitCode NodeMainInstance::Run() {
   CHECK_NOT_NULL(env);
 
   Context::Scope context_scope(env->context());
+  env.get()->InitializeEbpf();
+
```
Though this seems like a good location to do this, due how N|Solid is implemented, we should move it a bit earlier: to EnvList initialization.

santigimeno reviewed (changes requested):
In general I would change where the EbpfLoader is located and how it's implemented in terms of interface:
- EbpfLoader, unless I'm missing smthg, doesn't need to have static members nor API's.
- EbpfLoader should be a member of EnvList.
- EbpfLoader API should be exposed either by EnvList adding specific methods for loading, cleaning up, or even getting specific objects, or having a method that allows the user to grab the EbpfLoader instance so we can use it's methods.

santigimeno commented on src/nsolid/nsolid_api.cc:
```diff
@@ -2709,12 +2750,6 @@ static void DetectEBPFSupport(const FunctionCallbackInfo<Value>& args) {
-            OneByteString(isolate, "kernelVersion"),
-            v8::Integer::NewFromUnsigned(isolate, info.kernel_version))
-      .Check();
-
```
Why did this change ended up here?

RafaelGSS replied:
I think you are looking to an outdated version.

santigimeno commented on src/nsolid/nsolid_bpf.h:
```diff
@@ -3,6 +3,12 @@
 #if defined(NODE_WANT_INTERNALS) && NODE_WANT_INTERNALS
 
+// Forward declarations for BPF structures (outside the namespace)
+#if __has_include(<bpf/bpf.h>) && __has_include(<bpf/libbpf.h>)
+struct bpf_object;  // Forward declaration
+#define HAVE_LIBBPF 1
+#endif
+
```
Is this actually needed now that we embed libbpf?

RafaelGSS replied:
I think we should change the imports to relative `#include "bpf/bpf.h"` instead of `#include <bpf/bpf.h>` as I believe the former is getting from the system instead of deps.

In any case, I think we should have this define because libbpf will only be built on Linux, but if you try to build this on Windows/OSX, it will break without this flag.

santigimeno commented on src/nsolid/nsolid_bpf.h:
```diff
@@ -46,10 +54,36 @@ inline void cap_free(cap_t) {}
+#if HAVE_LIBBPF
+  std::vector<struct ::bpf_object*> loaded_objects_;
+#endif
+};
```
I still think we might need a mutex locking it. Use case: I want to access the bpf_object to actually communicate with the program, via any map defined by the program. It's not clear to me from which thread we will be accessing the vector.

RafaelGSS replied:
I'd prefer to add a mutex when we need that, for instance, when we create the eBPF program that makes use of this map. Wdyt?

santigimeno commented on deps/libbpf/libbpf-v1.5.1.tar.gz:
This needs to be removed. Or better: we need to create that dep updater script that will handle it.

RafaelGSS commented on src/nsolid/nsolid_bpf.cc:
```diff
@@ -0,0 +1,133 @@
+
+EbpfLoader::EbpfLoader()
+    : allowed_programs_({"hello_world"}),
+      ebpf_program_path_("/home/rafaelgss/repos/os/nsolid/src/ebpf/") {}
```
We should define where we are going to load the eBPF programs. How do we ship, and how will the users install it?

RafaelGSS committed:
deps: add libbpf dependency

RafaelGSS committed:
src: add eBPF program loader and initialization
This commit introduces a new initialization in
the startup queue: initializeEbpf()

The startup attempt to load into the kernel all ebpf
programs in the allowed_programs list.
